FPEngine.cc

```diff
@@ -470,17 +470,29 @@ void FPNetworkControl::probe_transmission_handler(nsock_pool nsp, nsock_event ns
         this->first_pcap_scheduled = true;
       }
 
-      buf = myprobe->getPacketBuffer(&len);
       /* Send the packet*/
-      assert(myprobe->host != NULL);
-      if (send_ip_packet(this->rawsd, myprobe->getEthernet(), myprobe->host->getTargetAddress(), buf, len) == -1) {
-        myprobe->setFailed();
-        this->cc_report_final_timeout();
-        myprobe->host->fail_one_probe();
-        gh_perror("Unable to send packet in %s", __func__);
+      for (int decoy = 0; decoy < o.numdecoys; decoy++) {
+        /* We don't need to change address if decoys aren't specified */
+        if (o.numdecoys > 1)
+          /* Decoys have to be sent only if changeSourceAddress worked */
+          assert(myprobe->changeSourceAddress(&((struct sockaddr_in6 *)&o.decoys[decoy])->sin6_addr) == OP_SUCCESS);
```

Comment:
Only use boolean checks in `assert`, not function calls with side-effects. `assert` is usually defined as a macro, so if asserts are turned off, the parameters may not be evaluated. Some implementations may evaluate the parameters more than once.